core/byzantine_test.go

```diff
@@ -0,0 +1,587 @@
+
+							idFn: node.addr,
+
+							buildProposalFn:          buildValidProposal,
+							buildPrePrepareMessageFn: node.buildPrePrepare,
+							buildPrepareMessageFn: func(_ []byte, view *proto.View) *proto.Message {
+								proposalHash := validProposalHash
+								if currentNode.byzantine {
+									proposalHash = []byte("invalid proposal hash")
+								}
+
+								return buildBasicPrepareMessage(
+									proposalHash,
+									currentNode.address,
+									view,
+								)
+							},
+							buildCommitMessageFn:      node.buildCommit,
+							buildRoundChangeMessageFn: node.buildRoundChange,
+
+							insertBlockFn: nil,
+							hasQuorumFn:   c.hasQuorumFn,
+						},
+
+						&mockTransport{multicastFn: c.gossip},
+					)
+				}
+			},
+		)
+
+		err := cluster.progressToHeight(10*time.Second, 1)
+		assert.NoError(t, err, "unable to reach height: %w", err)
+		assert.Equal(t, uint64(1), cluster.latestHeight)
+
+		cluster.makeNByzantine(int(cluster.maxFaulty()))
+		assert.NoError(t, cluster.progressToHeight(10*time.Second, 2))
+		assert.Equal(t, uint64(2), cluster.latestHeight)
+	})
+
+	t.Run("malicious +1 round in proposal", func(t *testing.T) {
+		t.Parallel()
+
+		cluster := newCluster(
+			6,
+			func(c *cluster) {
+				for _, node := range c.nodes {
+					currentNode := node
+					node.core = NewIBFT(
+						mockLogger{},
+						&mockBackend{
+							isValidBlockFn:         isValidProposal,
+							isValidProposalHashFn:  isValidProposalHash,
+							isValidSenderFn:        nil,
+							isValidCommittedSealFn: nil,
+							isProposerFn: func(from []byte, height uint64, round uint64) bool {
+								if round == 0 {
+									return false
+								}
+
+								return bytes.Equal(
+									from,
+									c.addresses()[int(round)%len(c.addresses())],
+								)
+							},
+
+							idFn: node.addr,
+
+							buildProposalFn: buildValidProposal,
+							buildPrePrepareMessageFn: func(
+								proposal []byte,
+								certificate *proto.RoundChangeCertificate,
+								view *proto.View,
+							) *proto.Message {
+								if currentNode.byzantine {
+									view.Round = uint64(rand.Int())
+								}
+
+								return buildBasicPreprepareMessage(
+									proposal,
+									validProposalHash,
+									certificate,
+									currentNode.address,
+									view,
+								)
+							},
+							buildPrepareMessageFn:     node.buildPrepare,
+							buildCommitMessageFn:      node.buildCommit,
+							buildRoundChangeMessageFn: node.buildRoundChange,
+
+							insertBlockFn: nil,
+							hasQuorumFn:   c.hasQuorumFn,
+						},
+
+						&mockTransport{multicastFn: c.gossip},
+					)
+				}
+			},
+		)
+
+		err := cluster.progressToHeight(20*time.Second, 1)
+		assert.NoError(t, err, "unable to reach height: %w", err)
+		assert.Equal(t, uint64(1), cluster.latestHeight)
+
+		// Max tolerant byzantine
+		cluster.makeNByzantine(int(cluster.maxFaulty()))
+		assert.NoError(t, cluster.progressToHeight(40*time.Second, 2))
```

Comment:
Nitpick
This timeout is the biggest value. (for example, 20sec or 30sec are used in other test) Is this value proper? Or is it possible to use same timeout and create function for test like?

```go
runTest := func(t *testing.T, cluster *cluster) {
  err := cluster.progressToHeight(20*time.Second, 1)
  assert.NoError(t, err, "unable to reach height: %w", err)
  assert.Equal(t, uint64(1), cluster.latestHeight)

  cluster.makeNByzantine(int(cluster.maxFaulty()))
  assert.NoError(t, cluster.progressToHeight(40*time.Second, 2))
}
```

Reply:
Some tests just take more time, so unfortunately no.